### PR TITLE
refactor: cache assets similar to how the normal compilation does it

### PR DESCRIPTION
**Pull request by uncleyo**
Previously `extract-text-webpack-plugin` emitted css assets every time anything _else_ changed.
Injected assets were also not included in compilation hash calculation and caused some rebuilds to be hidden.

There are no tests included in this PR, since this is watch-related and tests will go in a separate PR in the main repo that has great facilities for that.
This continues fixes started with https://github.com/webpack/webpack/pull/3737 and https://github.com/webpack/webpack/pull/3744 and should hopefully address a lot of watch related issues with `extract-text-webpack-plugin`.

### CODE REVIEW

**Comment by bebraw** [comment]
@d3viant0ne Any thoughts on this? Post final merge?

**Comment by michael-ciniawsky** [comment]
@uncleyo Please close and reopen the PR to trigger the CLA Bot again, sign the CLA and rebase against current master 😛

File: index.js [hunk 1]
```diff
@@ -309,7 +310,22 @@ ExtractTextPlugin.prototype.apply = function(compiler) {
 					}).replace(/\[(?:(\w+):)?contenthash(?::([a-z]+\d*))?(?::(\d+))?\]/ig, function() {
 						return loaderUtils.getHashDigest(source.source(), arguments[1], arguments[2], parseInt(arguments[3], 10));
 					});
+
+					var cacheKey = "extracted " + file;
+					var chunkHash = loaderUtils.getHashDigest(source.source());
+					if(compilation.cache && compilation.cache[cacheKey] && compilation.cache[cacheKey].hash === chunkHash) {
```
Comment (michael-ciniawsky) [inline index.js]: `\n`

**Comment by uncleyo** [comment]
_I've rebased this PR just to refresh it to latest master. I Don't think it should be merged._

For anyone interested, I've been using this plugin https://gist.github.com/uncleyo/106c3cb407bf5dd38ebd6be0710afd4a as a workaround for a while now. If anyone else is missing this feature, feel free to use it.

I don't like the code in this PR here myself, honestly, as I believe this is something webpack should handle automatically., either by exposing something like compilation.addAsset() or with plugin, something like the one in my gist.

I'll not close the PR just yet, hoping to hear some comments

**Comment by michael-ciniawsky** [comment]
@uncleyo Thx :) appeciated, well I'm not 💯 suited to give a meaningful answer about this I will try to bring @sokra in :D

**Comment by michael-ciniawsky** [comment]
@uncleyo Would you be interested in make the gist into a plugin `cache-webpack-plugin` or the like :)  ?  It's caching `assets` 'only' atm right ?

**Comment by michael-ciniawsky** [comment]
We are getting close to a new major release and therefore I will close this for now, feel free to open a new PR with these changes, when #540 landed on master, but as you commented this maybe better solved at core level in general 😛. Thx

**Pull request closed without merging**